In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import linear_model
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error
from sklearn.preprocessing import StandardScaler

In [2]:
customer = pd.read_csv('marketing_customer_analysis.csv')

# Lab | Customer Analysis Round 4

##### 1. Check the data types of the columns. Get the numeric data into dataframe called numerical and categorical columns in a dataframe called categoricals. (You can use np.number and np.object to select the numerical data types and categorical data types respectively)

In [3]:
# customer.dtypes

In [4]:
# cus_num = customer.select_dtypes(include = np.number)


In [5]:
# cus_cat = customer.select_dtypes(include = np.object)


##### 2. Now we will try to check the normality of the numerical variables visually
##### 2. 1 Use seaborn library to construct distribution plots for the numerical variables

In [6]:
# for col in cus_num:
#     sns.displot(cus_num[col])

##### 2.2 Use Matplotlib to construct histograms

In [7]:
# for col in cus_num:
    
#     plt.figure(figsize=(8, 5))  
    
#     plt.hist(cus_num[col], bins=20)
    
#     plt.title(f'Histogram of {col}')
#     plt.xlabel(col)
#     plt.ylabel('Frequency')
#     plt.show()

##### 2.3 Do the distributions for different numerical variables look like a normal distribution

In [8]:
# def log_transfom_clean(x):
#     x = np.log10(x)
#     if np.isfinite(x):
#         return x
#     else:
#         return 0

In [9]:
# cus_num = customer.select_dtypes(include = np.number)

# for col in cus_num:
#     plt.figure(figsize=(7, 4))
#     plt.hist(cus_num[col].apply(log_transfom_clean), bins=30)   
#     plt.title(f'Histogram of {col}')
#     plt.xlabel(col)
#     plt.ylabel('Frequency')
#     plt.show()

In [10]:
# def log_transform_clean2(x):
#     x = np.log10(x)
#     if np.isfinite(x):
#         return x
#     else:
#         return np.NAN

In [11]:
# cus_num = customer.select_dtypes(include = np.number)

# for col in cus_num:
    
#     cus_num[col] = list(map(log_transform_clean2, cus_num[col]))
#     cus_num[col] = cus_num[col].fillna(np.mean(cus_num[col]))
    
#     plt.figure(figsize=(7, 4))
#     plt.hist(cus_num[col].apply(log_transfom_clean), bins=30)   
#     plt.title(f'Histogram of {col}')
#     plt.xlabel(col)
#     plt.ylabel('Frequency')
#     plt.show()

# cus_num

I don´t think they could look as normal distributions using the log function. I could cut the values of the total claim amount at 600 but I wouldn´t do that because it provides lot of information. Here is the demonstration:

In [12]:
# cus_num = customer.select_dtypes(include = np.number)

# sns.boxplot(cus_num['Total Claim Amount'])
# plt.show()

In [13]:
# iqr = np.percentile(cus_num['Total Claim Amount'],75) - np.percentile(cus_num['Total Claim Amount'],25)

# upper_limit = np.percentile(cus_num['Total Claim Amount'],75) + 1.5*iqr
# lower_limit = np.percentile(cus_num['Total Claim Amount'],25) - 1.5*iqr

# print(upper_limit)
# print(lower_limit)

In [14]:
# cus_claim = cus_num[(cus_num['Total Claim Amount']>lower_limit)&(cus_num['Total Claim Amount']<2500)]
# sns.displot(cus_claim['Total Claim Amount'])
# plt.show()

In [15]:
# cus_claim['Total Claim Amount'] = list(map(log_transform_clean2, cus_claim['Total Claim Amount']))
# cus_claim['Total Claim Amount'] = cus_claim['Total Claim Amount'].fillna(np.mean(cus_claim['Total Claim Amount']))
# sns.displot(cus_claim['Total Claim Amount'])
# plt.show()

##### 3. For the numerical variables, check the multicollinearity between the features. Please note that we will use the column total_claim_amount later as the target variable.


In [16]:
# Reset data frame

# cus_num = customer.select_dtypes(include = np.number)

In [17]:
# Multicollinearity is a statistical phenomenon that occurs when two or more independent variables in a regression model are highly correlated with each other.

# We can create a matrix
# correlation_matrix = cus_num.corr()
# correlation_matrix


In [18]:
# Or a heatmap

# plt.figure(figsize=(8, 6))

# sns.heatmap(correlation_matrix, annot=True)
# plt.title('Correlation Matrix Heatmap')
# plt.show()

The values are not very correlated in general.

Looking at the Total Claim Amount, we can say that is quite correlated to the Monthly Premium Auto. The Customer Lifetime Value provides some information with a coefficient of correlation of 0.23.

Also, the income coefficient tells us about a negative correlation, which means that the Total Claim Amount gets higher when the Income amount goes lower.

##### 4. Drop one of the two features that show a high correlation between them (greater than 0.9). Write code for both the correlation matrix and for seaborn heatmap. If there is no pair of features that have a high correlation, then do not drop any features

I´m going to drop the values with very little correlation.

In [19]:
# cus_corr = cus_num.drop(['Months Since Last Claim', 'Months Since Policy Inception', 'Number of Open Complaints', 'Number of Policies'], axis=1)
# cus_corr

In [20]:
# correlation_matrix2 = cus_corr.corr()
# correlation_matrix2

In [21]:
# plt.figure(figsize=(4, 3))
# sns.heatmap(correlation_matrix2, annot=True)
# plt.title('Correlation Matrix Heatmap 2')
# plt.show()

# Lab | Customer Analysis Round 5

### Processing Data

##### 1. Further processing

I will compare the target value against the most correlated one, the Monthly Premium Auto.

In [22]:
# cus_num = customer.select_dtypes(include = np.number)

# y = cus_num['Total Claim Amount']
# X = cus_num[['Monthly Premium Auto']]

# lm = linear_model.LinearRegression()
# lm.fit(X,y)

# regression_line = lm.intercept_ + lm.coef_[0]*cus_num['Monthly Premium Auto']

# plt.plot(cus_num['Monthly Premium Auto'], regression_line, c = 'orange')
# sns.scatterplot(x='Monthly Premium Auto',y='Total Claim Amount',data=cus_num)
# plt.show()

I´ll check the R2-score

In [23]:
# print("R2-score is ", lm.score(X,y))
# print("mean squared error (MSE) is ", mean_squared_error(lm.predict(X),y))

The R2-socre indicates the model does not explain very well the variance in the target variable.
Also, we can see by the MSE that the model's predictions are not very close to the true values.

##### 2. X-y split.

Already done above

##### 3. Normalize (numerical).

In [24]:
y = customer['Total Claim Amount']
X = customer.drop(['Total Claim Amount'], axis=1)

cus_num = X.select_dtypes(include = np.number)
cus_cat = X.select_dtypes(include = object)

cus_tra = MinMaxScaler().fit(cus_num)
cus_nor = cus_tra.transform(cus_num)

cus_nor = pd.DataFrame(cus_nor, columns = cus_num.columns)

cus_nor.head()

Customer Lifetime Value    Income  Monthly Premium Auto  \
0                 0.010629  0.562847              0.033755   
1                 0.062406  0.000000              0.139241   
2                 0.134960  0.487763              0.198312   
3                 0.070589  0.000000              0.189873   
4                 0.011245  0.438443              0.050633   

   Months Since Last Claim  Months Since Policy Inception  \
0                 0.914286                       0.050505   
1                 0.371429                       0.424242   
2                 0.514286                       0.383838   
3                 0.514286                       0.656566   
4                 0.342857                       0.444444   

   Number of Open Complaints  Number of Policies  
0                        0.0               0.000  
1                        0.0               0.875  
2                        0.0               0.125  
3                        0.0               0.750  
4                        0.0               0.000

# Lab | Customer Analysis Round 6

### Processing Data - Deciding the Method

In [25]:
cus_cat.head()

Customer       State Response  Coverage Education Effective To Date  \
0  BU79786  Washington       No     Basic  Bachelor           2/24/11   
1  QZ44356     Arizona       No  Extended  Bachelor           1/31/11   
2  AI49188      Nevada       No   Premium  Bachelor           2/19/11   
3  WW63253  California       No     Basic  Bachelor           1/20/11   
4  HB64268  Washington       No     Basic  Bachelor            2/3/11   

  EmploymentStatus Gender Location Code Marital Status     Policy Type  \
0         Employed      F      Suburban        Married  Corporate Auto   
1       Unemployed      F      Suburban         Single   Personal Auto   
2         Employed      F      Suburban        Married   Personal Auto   
3       Unemployed      M      Suburban        Married  Corporate Auto   
4         Employed      M         Rural         Single   Personal Auto   

         Policy Renew Offer Type Sales Channel  Vehicle Class Vehicle Size  
0  Corporate L3           Offer1         Agent   Two-Door Car      Medsize  
1   Personal L3           Offer3         Agent  Four-Door Car      Medsize  
2   Personal L3           Offer1         Agent   Two-Door Car      Medsize  
3  Corporate L2           Offer1   Call Center            SUV      Medsize  
4   Personal L1           Offer1         Agent  Four-Door Car      Medsize

### Processing Data - OneHot Encoding

In [26]:
cus_one = cus_cat.drop(['Customer', 'Effective To Date'], axis=1)

cus_encoder = OneHotEncoder(drop='first').fit(cus_one)
cus_encoded = cus_encoder.transform(cus_one).toarray()

cus_encoded = pd.DataFrame(cus_encoded, columns=cus_encoder.get_feature_names_out())

cus_encoded.head()

State_California  State_Nevada  State_Oregon  State_Washington  \
0               0.0           0.0           0.0               1.0   
1               0.0           0.0           0.0               0.0   
2               0.0           1.0           0.0               0.0   
3               1.0           0.0           0.0               0.0   
4               0.0           0.0           0.0               1.0   

   Response_Yes  Coverage_Extended  Coverage_Premium  Education_College  \
0           0.0                0.0               0.0                0.0   
1           0.0                1.0               0.0                0.0   
2           0.0                0.0               1.0                0.0   
3           0.0                0.0               0.0                0.0   
4           0.0                0.0               0.0                0.0   

   Education_Doctor  Education_High School or Below  ...  \
0               0.0                             0.0  ...   
1               0.0                             0.0  ...   
2               0.0                             0.0  ...   
3               0.0                             0.0  ...   
4               0.0                             0.0  ...   

   Sales Channel_Branch  Sales Channel_Call Center  Sales Channel_Web  \
0                   0.0                        0.0                0.0   
1                   0.0                        0.0                0.0   
2                   0.0                        0.0                0.0   
3                   0.0                        1.0                0.0   
4                   0.0                        0.0                0.0   

   Vehicle Class_Luxury Car  Vehicle Class_Luxury SUV  Vehicle Class_SUV  \
0                       0.0                       0.0                0.0   
1                       0.0                       0.0                0.0   
2                       0.0                       0.0                0.0   
3                       0.0                       0.0                1.0   
4                       0.0                       0.0                0.0   

   Vehicle Class_Sports Car  Vehicle Class_Two-Door Car  Vehicle Size_Medsize  \
0                       0.0                         1.0                   1.0   
1                       0.0                         0.0                   1.0   
2                       0.0                         1.0                   1.0   
3                       0.0                         0.0                   1.0   
4                       0.0                         0.0                   1.0   

   Vehicle Size_Small  
0                 0.0  
1                 0.0  
2                 0.0  
3                 0.0  
4                 0.0  

[5 rows x 43 columns]

### Processing Data - Columns excluded in the model

I´m going to exclude the Customer identification and the effective date as I don´t consider them valuable information for the model.

### Processing Data - Concatenate

In [27]:
customer2 = pd.concat([cus_nor, cus_encoded], axis=1)
customer2

Customer Lifetime Value    Income  Monthly Premium Auto  \
0                    0.010629  0.562847              0.033755   
1                    0.062406  0.000000              0.139241   
2                    0.134960  0.487763              0.198312   
3                    0.070589  0.000000              0.189873   
4                    0.011245  0.438443              0.050633   
...                       ...       ...                   ...   
9129                 0.264137  0.719547              0.050633   
9130                 0.014719  0.216081              0.075949   
9131                 0.076951  0.000000              0.101266   
9132                 0.069098  0.219452              0.147679   
9133                 0.008766  0.000000              0.067511   

      Months Since Last Claim  Months Since Policy Inception  \
0                    0.914286                       0.050505   
1                    0.371429                       0.424242   
2                    0.514286                       0.383838   
3                    0.514286                       0.656566   
4                    0.342857                       0.444444   
...                       ...                            ...   
9129                 0.514286                       0.898990   
9130                 0.400000                       0.282828   
9131                 0.257143                       0.373737   
9132                 0.971429                       0.030303   
9133                 0.085714                       0.909091   

      Number of Open Complaints  Number of Policies  State_California  \
0                           0.0               0.000               0.0   
1                           0.0               0.875               0.0   
2                           0.0               0.125               0.0   
3                           0.0               0.750               1.0   
4                           0.0               0.000               0.0   
...                         ...                 ...               ...   
9129                        0.0               0.125               1.0   
9130                        0.0               0.000               1.0   
9131                        0.6               0.125               1.0   
9132                        0.0               0.250               1.0   
9133                        0.0               0.000               1.0   

      State_Nevada  State_Oregon  ...  Sales Channel_Branch  \
0              0.0           0.0  ...                   0.0   
1              0.0           0.0  ...                   0.0   
2              1.0           0.0  ...                   0.0   
3              0.0           0.0  ...                   0.0   
4              0.0           0.0  ...                   0.0   
...            ...           ...  ...                   ...   
9129           0.0           0.0  ...                   0.0   
9130           0.0           0.0  ...                   1.0   
9131           0.0           0.0  ...                   1.0   
9132           0.0           0.0  ...                   1.0   
9133           0.0           0.0  ...                   0.0   

      Sales Channel_Call Center  Sales Channel_Web  Vehicle Class_Luxury Car  \
0                           0.0                0.0                       0.0   
1                           0.0                0.0                       0.0   
2                           0.0                0.0                       0.0   
3                           1.0                0.0                       0.0   
4                           0.0                0.0                       0.0   
...                         ...                ...                       ...   
9129                        0.0                1.0                       0.0   
9130                        0.0                0.0                       0.0   
9131                        0.0                0.0                       0.0   
9132                  

### Linear Regression - Train-test split

In [28]:
y = customer['Total Claim Amount']
X = customer2
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

### Linear Regression - Apply linear regression

In [29]:
lm = linear_model.LinearRegression()
lm.fit(X_train,y_train)

predictions = lm.predict(X_train)
r2_score(y_train, predictions)

0.776198955247034

### Model Validation - R2

In [30]:
predictions_test = lm.predict(X_test)
r2_score(y_test, predictions_test)

0.7556097123007094

The model's predictions match the actual values of the dependent variable at a score of 0.76 which is not bad.

### Model Validation - MAE

In [31]:
mae = mean_absolute_error(y_test,predictions_test)
mae

97.74657243806372

In [32]:
customer.describe()

Customer Lifetime Value        Income  Monthly Premium Auto  \
count              9134.000000   9134.000000           9134.000000   
mean               8004.940475  37657.380009             93.219291   
std                6870.967608  30379.904734             34.407967   
min                1898.007675      0.000000             61.000000   
25%                3994.251794      0.000000             68.000000   
50%                5780.182197  33889.500000             83.000000   
75%                8962.167041  62320.000000            109.000000   
max               83325.381190  99981.000000            298.000000   

       Months Since Last Claim  Months Since Policy Inception  \
count              9134.000000                    9134.000000   
mean                 15.097000                      48.064594   
std                  10.073257                      27.905991   
min                   0.000000                       0.000000   
25%                   6.000000                      24.000000   
50%                  14.000000                      48.000000   
75%                  23.000000                      71.000000   
max                  35.000000                      99.000000   

       Number of Open Complaints  Number of Policies  Total Claim Amount  
count                9134.000000         9134.000000         9134.000000  
mean                    0.384388            2.966170          434.088794  
std                     0.910384            2.390182          290.500092  
min                     0.000000            1.000000            0.099007  
25%                     0.000000            1.000000          272.258244  
50%                     0.000000            2.000000          383.945434  
75%                     0.000000            4.000000          547.514839  
max                     5.000000            9.000000         2893.239678

The average difference between the predicted values and the actual values is 99. Considering the range 0 - 2893, the mean 434 and the 75% 547 I wouldn´t say is a bad result.

### Model Validation - MSE


In [33]:
predictions_test = lm.predict(X_test)
mse=mean_squared_error(y_test,predictions_test)
mse

19613.7026716937

### Model Validation - RMSE

In [34]:
rmse = np.sqrt(mean_squared_error(y_test,predictions_test))
rmse

140.04892956282708

Same comments than for MAE.